File: TwoSum.java
/**
 * https://leetcode.com/problems/two-sum/
 * @date 2020-08-05
 */
public class TwoSum {
    public static void main(String[] args) {
        int nums[] = {2, 0, 15, 7};
        int target = 9;
        System.out.printf("[%d, %d]\n", twoSum(nums, target)[0], twoSum(nums, target)[1]);
    }

    private static int[] twoSum(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (i != j && nums[i] + nums[j] == target) return new int[] {i , j};
            }
        }
        throw new IllegalArgumentException("No two sum solution");
    }
}

File: rust/two_sum.rs
//! <https://leetcode.com/problems/two-sum/>
use std::collections::HashMap;
pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut map = HashMap::new();
    for (i, num) in nums.into_iter().enumerate() {
        let key = target - num;
        if let Some(&index) = map.get(&key) {
            if i != index {
                return vec![index as _, i as _];
            }
        }
        map.insert(num, i);
    }
    vec![]
}

fn main() {
    assert_eq!(two_sum(vec![2, 7, 11, 15], 9), vec![0, 1]);
    assert_eq!(two_sum(vec![3, 2, 4], 6), vec![1, 2]);
    assert_eq!(two_sum(vec![3, 3], 6), vec![0, 1]);
}
